pyannote/core/interval_tree.py

```diff
@@ -25,286 +25,84 @@
-        for (segment1, segment2) in _co_iter(node1.left, node2.right):
-            yield segment1, segment2
-
-    # ----
-    # find segments of tree #2 intersecting current segment #1
-    # ----
-    segment1 = node1.key
-
-    # find segments of left tree #2 overlapping current segment #1
-    if node2.left:
-        for segment2 in _intersecting(node2.left, segment1):
-            yield segment1, segment2
-
-    # check if current segments #1 and #2 intersects
-    segment2 = node2.key
-    if segment1.intersects(segment2):
-        yield segment1, segment2
-
-    # find segments of right tree #2 overlapping current segment #1
-    if node2.right:
-        for segment2 in _intersecting(node2.right, segment1):
-            yield segment1, segment2
-
-    # ----
-    # if right tree #1 is not empty, process it
-    # ----
-    if node1.right:
-
-        for (segment1, segment2) in _co_iter(node1.right, node2.left):
-            yield segment1, segment2
-
-        segment2 = node2.key
-        for segment1 in _intersecting(node1.right, segment2):
-            yield segment1, segment2
-
-        for (segment1, segment2) in _co_iter(node1.right, node2.right):
-            yield segment1, segment2
-
-
-# =====================================================================
-# Banyan SortedSet updator
-# =====================================================================
-
-class TimelineUpdator(object):
-
-    class Metadata(object):
-
-        def update(self, key, key_fn, left, right):
-
-            # number of segments in tree
-            self.num = 1
-            # extent of tree
-            self.extent = key
-
-            if left:
-                self.num += left.num
-                self.extent = (min(self.extent[0], left.extent[0]),
-                               max(self.extent[1], left.extent[1]))
-            if right:
-                self.num += right.num
-                self.extent = (min(self.extent[0], right.extent[0]),
-                               max(self.extent[1], right.extent[1]))
-
-            self.extent = Segment(*self.extent)
-
-        def __repr__(self):
-            return '{extent: %s, num: %d}' % (repr(self.extent), self.num)
-
-    def length(self):
-        if self.root:
-            return self.root.metadata.num
-        else:
-            return 0
-
-    def kth(self, k):
-        """Get kth segment"""
-        if k < 0:
-            return _kth(self.root, self.root.num + k)
-        else:
-            return _kth(self.root, k)
-
-    def index(self, segment):
-        return _index(self.root, segment, 0)
+import collections as co
+import sortedcontainers as sc
+
+class SortedDict(sc.SortedDict):
+    length = dict.__len__
+
+class SortedSet(co.MutableSet):
+    "Sorted set intended only for segments."
+    def __init__(self, segments=()):
+        "Initialize sorted set from iterable of segments."
+        set_segments = set(segments)
+        self._set = set_segments
```

Comment:
I do it this way for three reasons:

1. It's an old C/C++-ism to declare your local variables at the top. I still think that way sometimes.
2. I think it's a bit more readable to see the local variable rather than "self._set" since the variable occurs again in the next two lines.
3. It's actually a bit faster this way (as opposed to referencing "self._set" in the next two lines). Nearly every operation in Python is an operator and even the dot, ".", as in "self._set" is an operator that becomes a function call and can be re-programmed. So by using the local we avoid those repeated function calls. Python's internal optimizer will propagate the local to all of its appearances but having "self._set" repeatedly will force a lookup. C-compilers often optimize the repeated lookups but Python cannot (unless you're using PyPy). This is definitely a micro-optimization which will have a small impact only in very hot code paths.